node/consensus/leader.go

```
@@ -306,47 +309,55 @@ func (ce *ConsensusEngine) processVotes(ctx context.Context) error {
+		ce.haltChan <- haltReason
+		return
+	}
 
-		votes := make([]*types.VoteInfo, 0)
-		for _, v := range ce.state.votes {
-			votes = append(votes, v)
-		}
+	if !ce.hasMajorityCeil(acks) {
+		// No majority yet, wait for more votes
+		ce.log.Info("Waiting for votes from the validators", "height", blkProp.height, "hash", blkProp.blkHash)
+		return
+	}
 
-		ci := &types.CommitInfo{
-			AppHash:      ce.state.blockRes.appHash,
-			Votes:        votes,
-			ParamUpdates: ce.state.blockRes.paramUpdates,
-		}
-		ce.state.commitInfo = ci
+	ce.log.Info("Majority of the validators have accepted the block, proceeding to commit the block",
+		"height", blkProp.blk.Header.Height, "hash", blkProp.blkHash, "acks", acks, "nacks", nacks)
 
-		// Commit the block and broadcast the blockAnn message
-		if err := ce.commit(ctx); err != nil {
-			ce.log.Errorf("Error committing the block (process votes): %v", err)
-			return err
+	votes := make([]*types.VoteInfo, 0, len(ce.state.votes))
+	for _, v := range ce.state.votes {
+		votes = append(votes, v)
+	}
+	slices.SortFunc(votes, func(a, b *types.VoteInfo) int {
+		if diff := bytes.Compare(a.Signature.PubKey, b.Signature.PubKey); diff != 0 {
+			return diff
 		}
+		return cmp.Compare(a.Signature.PubKeyType, b.Signature.PubKeyType)
+	})
+
+	// Set the commit info for the accepted block
+	ce.state.commitInfo = &types.CommitInfo{
+		AppHash:      blkRes.appHash,
+		Votes:        votes,
+		ParamUpdates: blkRes.paramUpdates,
+	}
 
-		ce.log.Infoln("Announce committed block", ce.state.blkProp.blk.Header.Height, ce.state.blkProp.blkHash,
-			ce.state.blockRes.paramUpdates)
-		// Broadcast the blockAnn message
-		go ce.blkAnnouncer(ctx, ce.state.blkProp.blk, ce.state.commitInfo)
+	// Commit the block and broadcast the blockAnn message
+	if err := ce.commit(ctx); err != nil {
+		ce.haltChan <- fmt.Sprintf("Error committing block %v: %v", blkProp.blkHash, err)
+		return
+	}
```

@charithabandi Is this the best way to handle this?  I noticed the `error` return from `processVotes` was unused and a `commit` error did not have an immediate effect.  Since all callers of `processVotes` ignored the error return, and there is a halt channel use in the `nack` majority case, I assume the intent is to halt via the channel if the commit failed.

---

Yeah , looks correct to me.